Update getting started (#43)
CDCgov/phdi-azure: docs/getting_started.md

Both the Python package and the containerized services can be used on their own in any manner users would like. However, the starter kit repositories combine the containerized services, with an orchestration layer (Azure Data Factory in this repository), CI/CD pipelines, and Infrastructure as Code to provide a "batteries included" cloud-based system for ingesting, standardizing, and enriching public health data. As documented in the [Implementation Guide](./implementation-guide.md)) the [Quick Start script](../quick-start.sh) automates the process of forking a starter kit repository into a repository owned by the user or their organization and deploying the entire pipeline to their own cloud environment. This high level architecture of the PHDI project is shown below.

![Architecture Diagram](./images/high-level-phdi-arch.png)

### Azure Starter Kit

> Source data is stored in Azure using [Azure Blob Storage](https://learn.microsoft.com/en-us/azure/storage/blobs/). When new data is written to this blob store the `read-source-data` [Azure Function](https://learn.microsoft.com/en-us/azure/data-factory/control-flow-azure-function-activity) is triggered. This function simply reads the new data from blob storage and passes it, along with some metadata, to the ingestion pipeline for further processing. The ingestion pipeline is implemented in [Azure Data Factory (ADF)](https://learn.microsoft.com/en-us/azure/data-factory/) and is responsible for data cleaning, standardization, enrichment, record linkage and finally uploading data to a FHIR server ([Azure API for FHIR](https://learn.microsoft.com/en-us/azure/healthcare-apis/azure-api-for-fhir/overview)). Each step of the pipeline, orchestrated by ADF, is simply a REST API call to indepent containerized microservices (AKA Building Blocks) developed by PHDI ([available here](https://github.com/orgs/CDCgov/packages?repo_name=phdi)). These services are deployed in Azure as [Azure Container Apps (ACA)](https://learn.microsoft.com/en-us/azure/container-apps/). Each endpoint on a service performs a single function, or step in a pipeline (e.g patient name standardization), and returns the processed data back to the data factory where it is passed on to the next service. Endpoints that perform related functions (e.g. name standardization and phone number standardization) are provided by a single service. Ingested data may be accessed via the FHIR server's API ([HL7 docs](https://hl7.org/fhir/http.html), [Azure docs](https://learn.microsoft.com/en-us/azure/healthcare-apis/azure-api-for-fhir/overview-of-search)), or the tabulation service. When called, the tabulation service with extract data, in tabular form according to a user-defined schema, from the FHIR server and writen to flat files. This entire architecture is show in the diagram below.

Reviewer: ```suggestion
Source data is stored in Azure using [Azure Blob Storage](https://learn.microsoft.com/en-us/azure/storage/blobs/). When new data is written to this blob store the `read-source-data` [Azure Function](https://learn.microsoft.com/en-us/azure/data-factory/control-flow-azure-function-activity) is triggered. This function simply reads the new data from blob storage and passes it, along with some metadata, to the ingestion pipeline for further processing. The ingestion pipeline is implemented in [Azure Data Factory (ADF)](https://learn.microsoft.com/en-us/azure/data-factory/) and is responsible for data cleaning, standardization, enrichment, record linkage, and finally uploading data to a FHIR server ([Azure API for FHIR](https://learn.microsoft.com/en-us/azure/healthcare-apis/azure-api-for-fhir/overview)). Each step of the pipeline, orchestrated by ADF, is simply a REST API call to independent containerized microservices (AKA Building Blocks) developed by PHDI ([available here](https://github.com/orgs/CDCgov/packages?repo_name=phdi)). These services are deployed in Azure as [Azure Container Apps (ACA)](https://learn.microsoft.com/en-us/azure/container-apps/). Each endpoint on a service performs a single function or step in a pipeline (e.g patient name standardization), and returns the processed data back to the data factor. It is then passed on to the next service. Endpoints that perform related functions (e.g. name standardization and phone number standardization) are provided by a single service. Ingested data may be accessed via the FHIR server's API ([HL7 docs](https://hl7.org/fhir/http.html), [Azure docs](https://learn.microsoft.com/en-us/azure/healthcare-apis/azure-api-for-fhir/overview-of-search)), or the tabulation service. When called, the tabulation service will extract data, in tabular form according to a user-defined schema, from the FHIR server and writen to flat files. This entire architecture is shown in the diagram below.
```

Reviewer: (Fixed a few typos and punctuation issues)

![Architecture Diagram](./images/azure-starter-kit-arch.drawio.png)

### Azure Blob Storage: Storage

All data for the starter kit in Azure is stored in [Azure Storage Accounts](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-overview), primarily in [Azure Blob Storage Containers](https://learn.microsoft.com/en-us/azure/storage/blobs/). Currently the starter kit includes the three storage accounts described below.

#### PHI (Protected Health Information) Storage Account

The PHI storage account is the primary storage account of the starter kit. All PHI/PII data stored outside of the FHIR server is stored in this account. The account has the following three containers:
1. `source-data`
2. `fhir-conversion-failures`
3. `fhir-upload-failures`

> Data in the form it is originally received by the starter kit is stored in the `source-data`. In practice the pipeline has two failure modes, on occasion issues will be encountered converting source data to FHIR at the beginning of the ingestion pipeline and uploading data to the FHIR server at the end of the ingestion pipeline. In the event of FHIR conversion and upload failures the data that could not be converted or uploaded is written to the appropriate container (`fhir-conversion-failures`, or `fhir-upload-failures`) as a JSON blob along with logging describing the specific failure. Additionally, each of these three containers contains a subdirectory for each type of public health data the starter kit can process. A new ELR message for the pipeline to ingest would be written to `source-data/elr/` while an eCR message that failed to be uploaded to the FHIR server would be written to `fhir-upload-failures/ecr`. The overall directory structure for the PHI storage account is shown below.

Reviewer: ```suggestion
Data in the form it is originally received by the starter kit is stored in the `source-data`. In practice the pipeline has two failure modes. On occasion, issues will be encountered converting source data to FHIR at the beginning of the ingestion pipeline and uploading data to the FHIR server at the end of the ingestion pipeline. In the event of FHIR conversion and upload failures, the data that could not be converted or uploaded is written to the appropriate container (`fhir-conversion-failures`, or `fhir-upload-failures`) as a JSON blob along with logging describing the specific failure. Additionally, each of these three containers contains a subdirectory for each type of public health data the starter kit can process. A new ELR message for the pipeline to ingest would be written to `source-data/elr/` while an eCR message that failed to be uploaded to the FHIR server would be written to `fhir-upload-failures/ecr`. The overall directory structure for the PHI storage account is shown below.
```

```bash
phdi{ENV NAME}phi{UNIX TIME OF CREATION}/
└── source-data/
    ├── elr/
    ├── ecr/
    └── vxu/
└── fhir-conversion-failures/
    ├── elr/
    ├── ecr/
    └── vxu/
└── fhir-upload-failures/
    ├── elr/
    ├── ecr/
    └── vxu/
```

#### Azure Functions Storage Account

> The second storage account in the starter kit is the Azure Functions Storage Account. This account is used for all storage needs related to Azure Functions. The starter kit currently creates a containers in this account to store the source code for the [`read-source-data](#azure-functions) function.

Reviewer: ```suggestion
The second storage account in the starter kit is the Azure Functions Storage Account. This account is used for all storage needs related to Azure Functions. The starter kit currently creates a container in this account to store the source code for the [`read-source-data](#azure-functions) function.
```

#### Terraform State Storage Account

> The third and final storage account in the starter kit in the Terraform State Storage Account. This account is shared across all deployed environments of the starter kit (dev, prod, QA, etc..) and is used to store the Terraform state of each environment. For more information Terraform state please refer to the [Infrastructure as Code (IaC)](#infrastructure-as-code-iac) section of this doc.

Reviewer: ```suggestion
The third and final storage account in the starter kit is the Terraform State Storage Account. This account is shared across all deployed environments of the starter kit (dev, prod, QA, etc..) and is used to store the Terraform state of each environment. For more information Terraform state please refer to the [Infrastructure as Code (IaC)](#infrastructure-as-code-iac) section of this doc.
```

### Azure Data Factory (ADF): Orchestration

> We use [Azure Data Factory (ADF)](https://learn.microsoft.com/en-us/azure/data-factory/control-flow-azure-function-activity) to define processes that require the use of multiple Building Blocks. These data flows can be developed using [The Azure Function Activity UI](https://learn.microsoft.com/en-us/azure/data-factory/control-flow-azure-function-activity#create-an-azure-function-activity-with-ui). However for best Infrastructure as Code (IaC) best practices we recommend deploying ADF pipelines in production from JSON configuration files. This allows consistency and reproducibility between deployments and the use of source control to track changes to the pipline over time. The JSON configuration file for the ingestion pipeline is available in [`/terraform/modules/data_factory/ingestion-pipeline.json`](../terraform/modules/data_factory/ingestion-pipeline.json) in this repository.

Reviewer 2: Double best: "However for best Infrastructure as Code (IaC) best practices we recommend"

Reviewer: Also `pipline` -> `pipeline`

| Service | Purpose | Documentation and Source Code |
| ------- | ------- | ----------------------------- |
| Fhir-converter | Convert Hl7v2 and C-CDA data to FHIR at the beginning of the ingestion pipeline. |  [https://github.com/CDCgov/phdi/tree/main/containers/fhir-converter](https://github.com/CDCgov/phdi/tree/main/containers/fhir-converter) |
| Ingestion | Provide endpoints for each step of the ingestion pipeline except conversion to FHIR. | [https://github.com/CDCgov/phdi/tree/main/containers/ingestion](https://github.com/CDCgov/phdi/tree/main/containers/ingestion) |
| Tabulation | Extract and tabularize data from the FHIR server according to user-defined schema. |  [https://github.com/CDCgov/phdi/tree/main/containers/tabulation](https://github.com/CDCgov/phdi/tree/main/containers/tabulation) |

### Azure Project Configuration

> Unknown - TODO??

Reviewer 2: This TODO meant to be left in?

Author: removed.
